### PR TITLE
config: catch all exceptions from YAML parser

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*:
YAML can throw a fairly wide variety of exceptions.  Unhandled exceptions are not handled well by the Envoy codebase in general, e.g. ThreadLocalStoreImpl will assert-fail if it's destructed right after it's constructed, which can happen with an exception.

*Risk Level*: Low

*Testing*:
//test/...
New unit test to cover type-conversion exception.

### CODE REVIEW

File: source/common/json/json_loader.cc [hunk 1]
```diff
@@ -743,6 +743,16 @@ ObjectSharedPtr Factory::loadFromYamlString(const std::string& yaml) {
...
+  } catch (YAML::BadConversion& e) {
+    throw EnvoyException(e.what());
+  } catch (...) {
+    // There is a potentially wide space of exceptions thrown by the YAML parser,
+    // and enumerating them all may be difficult. Envoy doesn't work well with
+    // unhandled exceptions, so we capture them and record the exception name in
+    // the Envoy Exception text.
+    std::exception_ptr p = std::current_exception();
+    throw EnvoyException(std::string("Unexpected YAML exception: ") +
+                         p.__cxa_exception_type()->name());
```
Comment (htuch) [inline source/common/json/json_loader.cc]: This `__cxa_exception_type` thing seems kind of scary; can you discuss how it works and any portability (compiler, libstdc++/libc++) concerns?

File: test/common/json/json_loader_test.cc [hunk 2]
```diff
@@ -445,5 +445,27 @@ TEST(JsonLoaderTest, YamlObject) {
...
+  bool exception_caught = false;
+  std::string exception_text;
+  try {
+    const Json::ObjectSharedPtr json = Json::Factory::loadFromYamlString(bad_yaml);
+  } catch (const EnvoyException& e) {
+    exception_caught = true;
+    exception_text = e.what();
+  }
+  EXPECT_TRUE(exception_caught);
+  EXPECT_TRUE(exception_text.find("bad conversion") != std::string::npos) << exception_text;
```
Comment (htuch) [inline test/common/json/json_loader_test.cc]: `EXPECT_THROW_WITH_REGEX`?

File: source/common/json/json_loader.cc [hunk 3]
```diff
@@ -743,6 +743,16 @@ ObjectSharedPtr Factory::loadFromYamlString(const std::string& yaml) {
     return parseYamlNode(YAML::Load(yaml));
   } catch (YAML::ParserException& e) {
     throw EnvoyException(e.what());
+  } catch (YAML::BadConversion& e) {
+    throw EnvoyException(e.what());
+  } catch (...) {
```
Comment (ggreenway) [inline source/common/json/json_loader.cc]: I'm against catching "...".  The things that it catches varies by platform, and it shouldn't be neccessary. It looks like everything from the YAML lib inherits from std::exception.  I say we catch that instead.

File: source/common/json/json_loader.cc (hunk 1 above)
Comment (jmarantz) [reply source/common/json/json_loader.cc]: fixed.

File: test/common/json/json_loader_test.cc (hunk 2 above)
Comment (jmarantz) [reply test/common/json/json_loader_test.cc]: I didn't find doc for that...but in any case I think that would probably be an adequate replacement for what I have now, but maybe not an adequate replacement for the negative test I just added, that shows we have explicitly looked for a BadConversion problem.

File: source/common/json/json_loader.cc [hunk 4]
```diff
@@ -743,6 +743,15 @@ ObjectSharedPtr Factory::loadFromYamlString(const std::string& yaml) {
...
   } catch (YAML::ParserException& e) {
     throw EnvoyException(e.what());
+  } catch (YAML::BadConversion& e) {
+    throw EnvoyException(e.what());
+  } catch (std::exception& e) {
+    // There is a potentially wide space of exceptions thrown by the YAML parser,
+    // and enumerating them all may be difficult. Envoy doesn't work well with
+    // unhandled exceptions, so we capture them and record the exception name in
+    // the Envoy Exception text.
+    std::exception_ptr p = std::current_exception();
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/json/json_loader.cc]: actually I didn't need that line at all.

File: source/common/json/json_loader.cc [hunk 5]
```diff
@@ -743,6 +743,15 @@ ObjectSharedPtr Factory::loadFromYamlString(const std::string& yaml) {
...
+    throw EnvoyException(std::string("Unexpected YAML exception: ") + e.what());
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/json/json_loader.cc]: Yes, there is no implicit conversion for the first arg of a +.

[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/json/json_loader.cc]: However I'm switching to using the formatter for consistency with the code below.

File: test/common/json/json_loader_test.cc (hunk 2 above)
Comment (jmarantz) [reply test/common/json/json_loader_test.cc]: It doesn't do *all* I want.  And I see what I couldn't find doc for it; it's defined in the Envoy codebase, so I added a parallel macro EXPECT_THROW_WITHOUT_REGEX and now I'm all set.

Having some trouble with docker for formatting fixups, and will push a new commit when that's sorted.

**Pull request merged**